Feature: improve order validation

vladimir-8 opened the pull request:
Fixes:

- https://github.com/coopcycle/coopcycle/issues/56
- https://github.com/coopcycle/coopcycle/issues/55

--

added: display validation errors on page load (address page & payment page)
added: validate order before processing the payment

added/updated: checkout end-to-end tests (includes: https://github.com/coopcycle/coopcycle-web/pull/4264 )

refactoring: move /checkout into /order (to match URL path)
refactoring: group payment-related components
refactoring: checkout log messages

--

How to test:

1. Create an order with a fixed time range (not ASAP)
2. Close a restaurant from the admin dashboard while you are on the different steps of the checkout flow (to simulate a situation when the time range is not valid any more)
3. See error message

r0xsh commented on js/app/components/Payment/index.js:
```diff
@@ -231,11 +134,56 @@ export default function(form, options) {
+
+      if (isGuest(options) || hasAccount) {
+        const validateItemRoute = window.Routing.generate("api_orders_validate_item", { id: options.orderId });
+        const result = await httpClient.get(validateItemRoute);
```
Nice use of the _auth client :upside_down_face: 

As describe in that [page](https://antoinebagnaud.notion.site/Inject-auth-into-Symfony-view-0977a318f79f48dca329d7760e6db2be) you can deconstruct the return to check if there is any error during the request

```javascript
const httpClient = new window._auth.httpClient();
const { response, error } = await httpClient.post(url, payload);

if (error)
{
	// Do some error handling stuff.
	console.error(error);
	return;
}

// Use response object as you want.
// The response is a parsed json.
console.log(response);
```
And i think a good default behavior is to add a generic error message that simply say "Something went wrong", it's not perfect but it's better than nothing for the user :)

https://github.com/coopcycle/coopcycle-web/blob/cfad1ed79f0b23d8cbd579be32eab5393cb2da4e/js/app/dashboard/components/TaskReportIncidentModalContent.js#L66-L73


The notification came from the [antd library](https://4x.ant.design/components/notification/), the modal is sexy and pretty handy to use

r0xsh commented on src/Security/OrderAccessTokenManager.php:
Nice :+1:

r0xsh reviewed (commented):
All i see is goodness :)
Nice addition on the logs side and some good refactoring along the way

vladimir-8 commented:
@r0xsh there were indeed no checks in the fulfilment method was enabled or not; added them:

https://github.com/coopcycle/coopcycle-web/pull/4263/commits/6643ee76e2dc861d2766e1338a7f415ef5b512b1#diff-837c1bf15cdb5fe35f25c02840a41f93bb4a0566909c4d0bd577e4d23d4c91c4R157-R164

https://github.com/coopcycle/coopcycle-web/pull/4263/commits/6643ee76e2dc861d2766e1338a7f415ef5b512b1#diff-5372d5dea5c96493ee69b31bf816e23c4b626051b578d8305da2c7d68765ab4fR75-R81

Also fixed https://github.com/coopcycle/coopcycle-app/issues/1734 in https://github.com/coopcycle/coopcycle-web/pull/4263/commits/ad46f76a5de84c335b46a24009ce84daa439d216

r0xsh commented:
Things seem to be working overall, but there's a small bug that occurs on the first action in the order process (as a client) after the restaurant timeslot is disabled. Nothing critical tho.
(Here is a video since it's a bit hard to describe the bug via text)

https://github.com/coopcycle/coopcycle-web/assets/5613584/4b51e9c5-512f-452c-a326-c09545e450f6

Here is the stacktrace 
```
Error:
Call to a member function getLower() on null

  at src/Validator/Constraints/ShippingTimeRangeJumpValidator.php:20
  at AppBundle\Validator\Constraints\ShippingTimeRangeJumpValidator->validate(array(object(TsRange), null), object(ShippingTimeRangeJump))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:759)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(array(object(TsRange), null), null, object(GenericMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:602)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateGenericNode(array(object(TsRange), null), null, null, object(GenericMetadata), '', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:117)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(array(object(TsRange), null), array(object(ShippingTimeRangeJump)), array('Default'))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveValidator.php:93)
  at Symfony\Component\Validator\Validator\RecursiveValidator->validate(array(object(TsRange), null), object(ShippingTimeRangeJump))
     (src/Form/Checkout/AbstractType.php:67)
  at AppBundle\Form\Checkout\AbstractType->AppBundle\Form\Checkout\{closure}('2024-06-05T16:00:00+02:00 - 2024-06-05T16:10:00+02:00', object(ExecutionContext), null)
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Constraints/CallbackValidator.php:37)
  at Symfony\Component\Validator\Constraints\CallbackValidator->validate('2024-06-05T16:00:00+02:00 - 2024-06-05T16:10:00+02:00', object(Callback))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:759)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup('2024-06-05T16:00:00+02:00 - 2024-06-05T16:10:00+02:00', null, object(GenericMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:602)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateGenericNode('2024-06-05T16:00:00+02:00 - 2024-06-05T16:10:00+02:00', object(Form), null, object(GenericMetadata), 'children[shippingTimeRange].data', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:117)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate('2024-06-05T16:00:00+02:00 - 2024-06-05T16:10:00+02:00', array(object(Callback)), array('Default'))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Validator/Constraints/FormValidator.php:138)
  at Symfony\Component\Form\Extension\Validator\Constraints\FormValidator->validate(object(Form), object(Form))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:759)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Form), '0000000000001c660000000000000000', object(GenericMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:602)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateGenericNode(object(Form), object(Form), '0000000000001c660000000000000000', object(GenericMetadata), 'children[shippingTimeRange]', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:117)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Form), array(object(Form)))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Validator/Constraints/FormValidator.php:145)
  at Symfony\Component\Form\Extension\Validator\Constraints\FormValidator->validate(object(Form), object(Form))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:759)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateInGroup(object(Form), '0000000000001bdc0000000000000000', object(ClassMetadata), 'Default', object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:488)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateClassNode(object(Form), '0000000000001bdc0000000000000000', object(ClassMetadata), '', array('Default'), null, 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:313)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validateObject(object(Form), '', array('Default'), 1, object(ExecutionContext))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveContextualValidator.php:138)
  at Symfony\Component\Validator\Validator\RecursiveContextualValidator->validate(object(Form), null, array('Default'))
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/RecursiveValidator.php:93)
  at Symfony\Component\Validator\Validator\RecursiveValidator->validate(object(Form), null, null)
     (vendor/symfony/symfony/src/Symfony/Component/Validator/Validator/TraceableValidator.php:66)
  at Symfony\Component\Validator\Validator\TraceableValidator->validate(object(Form))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php:50)
  at Symfony\Component\Form\Extension\Validator\EventListener\ValidationListener->validateForm(object(PostSubmitEvent), 'form.post_submit', object(EventDispatcher))
     (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:230)
  at Symfony\Component\EventDispatcher\EventDispatcher->callListeners(array(object(Closure), object(Closure)), 'form.post_submit', object(PostSubmitEvent))
     (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:59)
  at Symfony\Component\EventDispatcher\EventDispatcher->dispatch(object(PostSubmitEvent), 'form.post_submit')
     (vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ImmutableEventDispatcher.php:33)
  at Symfony\Component\EventDispatcher\ImmutableEventDispatcher->dispatch(object(PostSubmitEvent), 'form.post_submit')
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:681)
  at Symfony\Component\Form\Form->submit(array(), true)
     (vendor/sylius/resource-bundle/src/Bundle/Form/Extension/HttpFoundation/HttpFoundationRequestHandler.php:107)
  at Sylius\Bundle\ResourceBundle\Form\Extension\HttpFoundation\HttpFoundationRequestHandler->handleRequest(object(Form), object(Request))
     (vendor/symfony/symfony/src/Symfony/Component/Form/Form.php:503)
  at Symfony\Component\Form\Form->handleRequest(object(Request))
     (src/Controller/OrderController.php:223)
  at AppBundle\Controller\OrderController->indexAction(object(Request), object(OrderManager), object(CompositeCartContext), object(CompositeOrderProcessor), object(DataCollectorTranslator), object(SettingsManager), object(Context), object(Session))
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:163)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/HttpKernel.php:75)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/symfony/src/Symfony/Component/HttpKernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (web/index.php:28)                
```

r0xsh reviewed (approved):
LGTM, nice work :)

r0xsh commented:
I guess you have to fix the tests, but the feature itself work as expected :+1:
